LocalApi.py

```diff
@@ -0,0 +1,250 @@
+import random
+import logging
+import bz2
+import pickle
+import itertools
+import xml.etree.ElementTree as xml
+import mwparserfromhell
+from titlecase import titlecase
+from os.path import getsize
+from IWikiApi import IWikiApi
+
+log = logging.getLogger(__name__)
+
+class WikipediaIndexFile:
+    """
+    A class to represent the index file provided with the data dump
+    The index file is a sequence of lines, each of the form:
+    BZ2_OFFSET:ARTICLE_ID:ARTICLE_TITLE
+
+    When we read the index file, we want to keep track of the ARTICLE_TITLE and the BZ2_OFFSET.
+    Articles are arranged in groups with the same BZ2_OFFSET,
+    and we need to know the BZ2_OFFSET of the group and the BZ2_OFFSET of the group after.
+
+    This allows us to take a clipping of the BZ2 file to decompress.
+    """
+    def __init__(self, index_file, xml_file_size):
+        self.index_file = index_file
+        self.xml_file_size = xml_file_size
+        self.articles = {}
+        self.article_titles = []
```

Comment:
A `set()` would have faster lookup times

Reply:
You can't do `random.choice` on a set:

```
random.choice(set(a.keys()))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/random.py", line 261, in choice
    return seq[i]
TypeError: 'set' object does not support indexing
```